src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,384 @@
+                        }
+                        Update(update) => {
+                            let tx = tx.as_mut().ok_or_else(|| {
+                                anyhow!("got row event without a prior BEGIN event")
+                            })?;
+                            let old_row = self.row_from_tuple(update.old_tuple().unwrap())?;
+                            tx.delete(old_row).await?;
+
+                            let new_row = self.row_from_tuple(update.new_tuple())?;
+                            tx.insert(new_row).await?;
+                        }
+                        Delete(delete) => {
+                            let tx = tx.as_mut().ok_or_else(|| {
+                                anyhow!("got row event without a prior BEGIN event")
+                            })?;
+                            let row = self.row_from_tuple(delete.old_tuple().unwrap())?;
+                            tx.delete(row).await?;
+                        }
+                        Commit(commit) => {
+                            self.lsn = commit.commit_lsn().into();
+                            // Release the lease
+                            tx = None;
+                        }
+                        Origin(_) | Relation(_) | Type(_) => (),
+                        Truncate(_) => bail!("source table got truncated"),
+                        // The enum is marked as non_exaustive. Better to be conservative here in
+                        // case a new message is relevant to the semantics of our source
+                        _ => bail!("unexpected logical replication message"),
+                    }
+                }
+                // The upstream will periodically send keepalive messages to:
+                //    a) keep the TCP connection alive
+                //    b) get feedback about the state of replication in the replica
+                // The upstream database can dictate if a reply MUST be given by setting the reply
+                // field to 1
+                Ok(ReplicationMessage::PrimaryKeepAlive(keepalive)) => {
+                    if keepalive.reply() == 1 {
+                        let ts: i64 = PG_EPOCH
+                            .elapsed()
+                            .expect("system clock set earlier than year 2000!")
+                            .as_micros()
+                            .try_into()
+                            .expect("software more than 200k years old, consider updating");
+
+                        // Here we report that we've seen everything up until `self.lsn`, which is
+                        // the lsn of that commit message we received. However, there could be
+                        // transactions that we have not seen yet (perhaps because they are not
+                        // committed yet) that started way earlier than self.lsn. Therefore, it's
+                        // unclear how to calculate a safe lower bound value to report back to the
```

Comment:
Yes, this is necessary for launching yes. But fortunately it consists of figuring out how pg itself does it because it must be doing it correctly. I'll address this in a future PR